File: apps/mobile/lib/features/fuel/data/fuel_queue_repository.dart
import 'package:drift/drift.dart';
import 'package:uuid/uuid.dart';

import '../../../core/database/app_database.dart';

/// File d'attente locale des déclarations de plein ([PendingFuelRecords]) —
/// alimentée par `FuelFlowNotifier` quand une tentative de soumission échoue
/// pour une raison réseau (pas de connexion), rejouée une par une par
/// `SyncService` via [FuelRepository] — même pattern que
/// `ValidationQueueRepository` (`lib/features/qr/data/validation_queue_repository.dart`).
class FuelQueueRepository {
  FuelQueueRepository({required AppDatabase database}) : _db = database;

  final AppDatabase _db;
  final _uuid = const Uuid();

  /// Met une déclaration en attente et renvoie le `clientEventId` généré,
  /// afin qu'un futur envoi (immédiat ou différé) soit reconnu comme le même
  /// événement côté serveur.
  Future<String> enqueue({
    required String vehicleId,
    required double liters,
    required double totalCost,
    required double odometer,
    required String fuelType,
    String? stationName,
    required double latitude,
    required double longitude,
    required DateTime recordedAt,
    required String receiptPhotoPath,
    required String odometerPhotoPath,
  }) async {
    final clientEventId = _uuid.v4();
    await _db.into(_db.pendingFuelRecords).insert(
          PendingFuelRecordsCompanion.insert(
            clientEventId: clientEventId,
            vehicleId: vehicleId,
            liters: liters,
            totalCost: totalCost,
            odometer: odometer,
            fuelType: fuelType,
            stationName: Value(stationName),
            latitude: latitude,
            longitude: longitude,
            recordedAt: recordedAt,
            receiptPhotoPath: receiptPhotoPath,
            odometerPhotoPath: odometerPhotoPath,
          ),
        );
    return clientEventId;
  }

  /// Nombre de déclarations encore `pending`/`failed` — utilisé par l'écran
  /// de synchronisation.
  Stream<int> watchPendingCount() {
    final query = _db.selectOnly(_db.pendingFuelRecords)
      ..addColumns([_db.pendingFuelRecords.id.count()])
      ..where(
        _db.pendingFuelRecords.syncStatus.equalsValue(SyncStatus.pending) |
            _db.pendingFuelRecords.syncStatus.equalsValue(SyncStatus.failed),
      );
    return query
        .map((row) => row.read(_db.pendingFuelRecords.id.count()) ?? 0)
        .watchSingle();
  }

  Stream<int> watchFailedCount() {
    final query = _db.selectOnly(_db.pendingFuelRecords)
      ..addColumns([_db.pendingFuelRecords.id.count()])
      ..where(_db.pendingFuelRecords.syncStatus.equalsValue(SyncStatus.failed));
    return query
        .map((row) => row.read(_db.pendingFuelRecords.id.count()) ?? 0)
        .watchSingle();
  }

  /// Déclarations prêtes à être envoyées — une par une (upload multipart),
  /// même règles de backoff/forçage que `ValidationQueueRepository.nextBatch`.
  Future<List<PendingFuelRecord>> nextBatch({
    required int limit,
    required bool force,
    required int maxAutoRetries,
  }) {
    final now = DateTime.now();
    final query = _db.select(_db.pendingFuelRecords)
      ..where((t) {
        var predicate = t.syncStatus.equalsValue(SyncStatus.pending) |
            t.syncStatus.equalsValue(SyncStatus.failed);
        if (!force) {
          predicate = predicate & t.retryCount.isSmallerThanValue(maxAutoRetries);
          predicate = predicate &
              (t.nextRetryAt.isNull() | t.nextRetryAt.isSmallerOrEqualValue(now));
        }
        return predicate;
      })
      ..orderBy([(t) => OrderingTerm.asc(t.recordedAt)])
      ..limit(limit);
    return query.get();
  }

  Future<void> markUploading(int id) {
    return (_db.update(_db.pendingFuelRecords)..where((t) => t.id.equals(id)))
        .write(
      const PendingFuelRecordsCompanion(
        syncStatus: Value(SyncStatus.uploading),
      ),
    );
  }

  Future<void> markSynced(int id) {
    return (_db.update(_db.pendingFuelRecords)..where((t) => t.id.equals(id)))
        .write(
      const PendingFuelRecordsCompanion(
        syncStatus: Value(SyncStatus.synced),
        lastError: Value(null),
        nextRetryAt: Value(null),
      ),
    );
  }

  Future<void> markFailed({
    required int id,
    required int previousRetryCount,
    String? error,
  }) {
    final retryCount = previousRetryCount + 1;
    return (_db.update(_db.pendingFuelRecords)..where((t) => t.id.equals(id)))
        .write(
      PendingFuelRecordsCompanion(
        syncStatus: const Value(SyncStatus.failed),
        retryCount: Value(retryCount),
        lastError: Value(error),
        nextRetryAt: Value(_nextRetryAt(retryCount)),
      ),
    );
  }

  Future<void> resetStaleUploading() {
    return (_db.update(_db.pendingFuelRecords)
          ..where((t) => t.syncStatus.equalsValue(SyncStatus.uploading)))
        .write(
      const PendingFuelRecordsCompanion(syncStatus: Value(SyncStatus.pending)),
    );
  }
}

/// Backoff exponentiel partagé avec les autres files : 5s, 10s, 20s, 40s...
/// plafonné à 5 minutes.
DateTime _nextRetryAt(int retryCount) {
  final seconds = (5 * (1 << retryCount.clamp(0, 10))).clamp(5, 300);
  return DateTime.now().add(Duration(seconds: seconds));
}

File: apps/mobile/src/database/fuel-queue.repository.ts
import { getDatabase } from './db';
import { PendingFuelRecordRow } from '../types/sync.types';
import { CreateFuelRecordMetadataDto, FuelType } from '../types/fuel.types';

export interface EnqueueFuelParams {
  clientEventId?: string;
  vehicleId: string;
  liters: number;
  totalCost: number;
  odometer: number;
  fuelType: FuelType;
  stationName?: string | null;
  latitude: number;
  longitude: number;
  recordedAt?: string;
  receiptPhotoPath: string;
  odometerPhotoPath: string;
}

export const FuelQueueRepository = {
  enqueue(params: EnqueueFuelParams): string {
    const db = getDatabase();
    const clientEventId = params.clientEventId || `fuel_${Date.now()}_${Math.random().toString(36).substring(2, 9)}`;
    const recordedAt = params.recordedAt || new Date().toISOString();
    const createdAtDevice = new Date().toISOString();

    db.runSync(
      `INSERT OR IGNORE INTO pending_fuel_records (
        client_event_id, vehicle_id, liters, total_cost, odometer,
        fuel_type, station_name, latitude, longitude, recorded_at,
        receipt_photo_path, odometer_photo_path, created_at_device,
        sync_status, retry_count
      ) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, 'pending', 0)`,
      [
        clientEventId,
        params.vehicleId,
        params.liters,
        params.totalCost,
        params.odometer,
        params.fuelType,
        params.stationName || null,
        params.latitude,
        params.longitude,
        recordedAt,
        params.receiptPhotoPath,
        params.odometerPhotoPath,
        createdAtDevice,
      ]
    );

    return clientEventId;
  },

  getNextBatch(limit = 5, force = false, maxAutoRetries = 8): PendingFuelRecordRow[] {
    const db = getDatabase();
    const nowIso = new Date().toISOString();

    if (force) {
      return db.getAllSync<PendingFuelRecordRow>(
        `SELECT * FROM pending_fuel_records 
         WHERE sync_status != 'synced' AND sync_status != 'uploading'
         ORDER BY recorded_at ASC LIMIT ?`,
        [limit]
      );
    }

    return db.getAllSync<PendingFuelRecordRow>(
      `SELECT * FROM pending_fuel_records 
       WHERE sync_status IN ('pending', 'failed')
         AND retry_count < ?
         AND (next_retry_at IS NULL OR next_retry_at <= ?)
       ORDER BY recorded_at ASC LIMIT ?`,
      [maxAutoRetries, nowIso, limit]
    );
  },

  markUploading(id: number): void {
    const db = getDatabase();
    db.runSync(`UPDATE pending_fuel_records SET sync_status = 'uploading' WHERE id = ?`, [id]);
  },

  markSynced(id: number): void {
    const db = getDatabase();
    db.runSync(`DELETE FROM pending_fuel_records WHERE id = ?`, [id]);
  },

  markFailed(id: number, previousRetryCount: number, error: string): void {
    const db = getDatabase();
    const nextRetryCount = previousRetryCount + 1;
    const delaySeconds = Math.min(300, 5 * Math.pow(2, Math.max(0, nextRetryCount - 1)));
    const nextRetryAt = new Date(Date.now() + delaySeconds * 1000).toISOString();

    db.runSync(
      `UPDATE pending_fuel_records 
       SET sync_status = 'failed', retry_count = ?, last_error = ?, next_retry_at = ?
       WHERE id = ?`,
      [nextRetryCount, error, nextRetryAt, id]
    );
  },

  resetStaleUploading(): void {
    const db = getDatabase();
    db.runSync(`UPDATE pending_fuel_records SET sync_status = 'pending' WHERE sync_status = 'uploading'`);
  },

  getPendingCount(): number {
    const db = getDatabase();
    const row = db.getFirstSync<{ count: number }>(
      `SELECT COUNT(*) as count FROM pending_fuel_records WHERE sync_status != 'synced'`
    );
    return row?.count || 0;
  },

  getFailedCount(): number {
    const db = getDatabase();
    const row = db.getFirstSync<{ count: number }>(
      `SELECT COUNT(*) as count FROM pending_fuel_records WHERE sync_status = 'failed'`
    );
    return row?.count || 0;
  },

  toMetadata(row: PendingFuelRecordRow): CreateFuelRecordMetadataDto {
    return {
      clientEventId: row.client_event_id,
      vehicleId: row.vehicle_id,
      liters: row.liters,
      totalCost: row.total_cost,
      odometer: row.odometer,
      fuelType: row.fuel_type as FuelType,
      stationName: row.station_name || undefined,
      latitude: row.latitude,
      longitude: row.longitude,
    };
  },
};
